# Lab 4 - Information Theory in Machine Learning

Welcome to this week's lab on Information Theory! This week, we will dive into the fascinating world of Information Theory as applied to Machine Learning. Specifically, we will focus on two key concepts: Entropy and Information Gain. These principles are fundamental in understanding how decision trees make split decisions to organize data effectively.

### Entropy
- Entropy, in the context of information theory, measures the level of uncertainty or disorder within a set of data.
- In machine learning, particularly in decision trees, entropy helps to determine how a dataset should be split. A high entropy means more disorder, indicating that our dataset is varied. Conversely, low entropy suggests more uniformity in the data.

### Information Gain
- Information Gain measures the reduction in entropy after the dataset is split on an attribute.
- It is crucial in building decision trees as it helps to decide the order of attributes the tree will use for splitting the data. The attribute with the highest Information Gain is chosen as the splitting attribute at each node.

## Part 1: Entropy and Information Gain in Decision Trees
Decision Trees use these concepts to create branches. By choosing splits that maximize Information Gain (or equivalently minimize entropy), a decision tree can effectively categorize data, leading to better classification or regression models.

### Step 1: Import Necessary Libraries

In [6]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree
import matplotlib.pyplot as plt

### Step 2: Load and Explore the Iris Dataset

In [7]:
iris = load_iris()
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df['target'] = iris.target

### Step 3: Calculate Entropy
To calculate the `entropy` we need to:
- First, extract the target variable `y` from your dataset (like the 'target' column in the Iris dataset).
- Then, call `calculate_entropy(y)` to get the entropy.

This function calculates the entropy of a given target variable `y`. It works by first determining the unique classes in `y`, then computes the probability of each class, and uses this probability to calculate the entropy. This is a crucial step in understanding the disorder or uncertainty in the dataset, a fundamental concept in information theory.

In [8]:
def calculate_entropy(y):
    class_labels = np.unique(y)
    entropy = 0
    for label in class_labels:
        probability = len(y[y == label]) / len(y)
        entropy -= probability * np.log2(probability)
    return entropy

target_entropy = calculate_entropy(df['target'])
print('Target variable Entropy:', target_entropy)

Target variable Entropy: 1.584962500721156


What is your observastion about the calculated Entropy?

We got an entropy value of 1.584962500721156 which indicates that we have a mix of the different groups in the data. This makes it neither organized nor messy, and it could be a good variety for decision making models.

### Step 4: Calculate Information Gain
There are three steps for calculating the Information Gain:
1. Compute Overall Entropy: Use the entropy function from Step 3 on the entire target dataset.
2. Calculate Weighted Entropy for Each Attribute: For each unique value in the attribute, partition the dataset and calculate its entropy. Then calculate the weighted sum of these entropies, where the weights are the proportions of instances in each partition.
3. Compute Information Gain: Subtract the weighted entropy of the split from the original entropy.

The attribute with the highest Information Gain is generally chosen for splitting, as it provides the most significant reduction in uncertainty. This step is critical in constructing an effective decision tree, as it directly influences the structure and depth of the tree.

In [9]:
def calculate_information_gain(df, attribute, target_name):
    total_entropy = calculate_entropy(df[target_name])
    values, counts = np.unique(df[attribute], return_counts=True)
    weighted_entropy = sum((counts[i] / sum(counts)) * calculate_entropy(df.where(df[attribute] == values[i]).dropna()[target_name]) for i in range(len(values)))
    information_gain = total_entropy - weighted_entropy
    return information_gain

print('Information Gain:')
for column in df.columns[:-1]:
    print (column)
    print(calculate_information_gain(df, column, 'target'))

Information Gain:
sepal length (cm)
0.8769376208910578
sepal width (cm)
0.5166428756804977
petal length (cm)
1.4463165236458
petal width (cm)
1.4358978386754417


Discuss your findings here.

We can see from the results we obtained that sepal length and sepal width give moderate and lower information gains respectively. Petal length and petal width give high information gain, and finally the target gives the highest information gain. 

This allows us to conclude that the target variable, along with petal length and petal width are important for effective decision tree splits. Sepal length and sepal width do not provide enough information for splitting.

## Part 2: Apply Entropy and Information Gain on a different dataset

Your task is to choose a new dataset and implement what you learned in `Part 1` on this new dataset.

### Task 1: Implement Entropy and Information Gain

In [10]:
# Your code goes here
import seaborn as sns
tips = sns.load_dataset('tips')

target_variable = 'tip'

for column in tips.select_dtypes(include='number'):
    information_gain_result = calculate_information_gain(tips, column, target_variable)
    print(f"Attribute: {column}, Information Gain: {information_gain_result}")

Attribute: total_bill, Information Gain: 5.835293258189954
Attribute: tip, Information Gain: 5.944944436477509
Attribute: size, Information Gain: 0.944204825924154


### Task 2: Discuss your findings in detail
Provide detailed explanation and discussion about your findings.

Regarding the chosen dataset 'tips', we get the highest infromation from the target variable. However, for both tip and total_bill, we get an information gain superior to 1 and therefore it is too much information. The model training would be too complex and we would need a more complicated classifier. On the other hand, the size variable has less information gain while still being highly informative. 

## Submission
Submit a link to your completed Jupyter Notebook file hosted on your private GitHub repository through the submission link in Blackboard.